### PR TITLE
gtkextra 3.1.2 (new formula)

### PR DESCRIPTION
**Pull request by tschoonj**
Hi,

I created this formula for gtkextra, a library with additional widgets for Gtk2. 
Two remarks:

1. I wasn't entirely sure whether to contribute here or at homebrew-x11. It does not depend on x11 directly, only through gtk+. However, if gtk+ is built with quartz backend, gtkextra will build just as well. My only worry here is that somehow gtkextra will need to get rebuilt if the user decides to change the backend of gtk+.
2. The test block: since this is part of a GUI toolkit it didn't make a lot of sense to me including a test here so I just changed the default line to: `system "true"`. Hope this is ok

Best,

Tom

### CODE REVIEW

File: Library/Formula/gtkextra.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+                          "--disable-dependency-tracking",
+                          "--disable-silent-rules",
+                          "--enable-tests",
+                          "--prefix=#{prefix}"
+    system "make"
+    system "make", "install"
+  end
+
+  test do
+    system "true"
```
Comment (mikemcquaid) [inline Library/Formula/gtkextra.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

**Comment by tschoonj** [comment]
Hi,

This is a library that provides no executable. I could compile a short example that links against the library but I won't be able to run it really since it will be a GUI. 

What do you think?

**Comment by mikemcquaid** [comment]
Are there any library functions you can call that don't spawn a GUI?

**Comment by tschoonj** [comment]
nope... this whole library just offers additional gtk widgets

**Comment by mikemcquaid** [comment]
Ok, just remove the `test do` block then, thanks.

**Comment by mikemcquaid** [comment]
Although it looks like you could probably e.g. create a sheet and just not show the window: http://gtkextra.sourceforge.net/cms/index.php?option=com_wrapper&view=wrapper&Itemid=11

**Comment by tschoonj** [comment]
Hi I am trying to write a test but I am running in trouble with the compilation command. I am using this:

````
system ENV.cc, "-o", "test", "`pkg-config", "--cflags", "gtkextra-3.0`", "test.c", "`pkg-config", "--libs", "gtkextra-3.0`"
````

The following is then reported:

````
Testing gtkextra
==> /usr/bin/clang -o test `pkg-config --cflags gtkextra-3.0` test.c `pkg-config --libs gtkextra-3.0`
clang: error: unsupported option '--libs'
clang: error: no such file or directory: '`pkg-config'
clang: error: no such file or directory: 'gtkextra-3.0`'
clang: error: no such file or directory: '`pkg-config'
clang: error: no such file or directory: 'gtkextra-3.0`'
Error: gtkextra: failed
Failed executing: /usr/bin/clang -o test `pkg-config --cflags gtkextra-3.0` test.c `pkg-config --libs gtkextra-3.0`
/usr/local/Library/Homebrew/formula.rb:703:in `block in system'
/usr/local/Library/Homebrew/formula.rb:660:in `open'
/usr/local/Library/Homebrew/formula.rb:660:in `system'
/usr/local/Library/Formula/gtkextra.rb:279:in `block in <class:Gtkextra>'
/usr/local/Library/Homebrew/formula.rb:613:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:21:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:610:in `run_test'
/usr/local/Library/Homebrew/cmd/test.rb:59:in `block (2 levels) in test'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/cmd/test.rb:58:in `block in test'
/usr/local/Library/Homebrew/cmd/test.rb:37:in `each'
/usr/local/Library/Homebrew/cmd/test.rb:37:in `test'
/usr/local/Library/brew.rb:135:in `<main>'
````

How do I get ruby's system command to pass these backticks properly?

**Comment by mikemcquaid** [comment]
You can't with the system command. You can just hardcode the values output by `pkg-config`

**Comment by tschoonj** [comment]
that won't work: the name of the gtkextra library depends on whether gtk was built with the x11 or quartz backend.

Something like this seems to work better though:

````
cflags = `pkg-config --cflags gtkextra-3.0`
libs = `pkg-config --libs gtkextra-3.0`
system ENV.cc, "-o", "test", cflags , "test.c", libs
system "./test"
````

However, then I get the following error messages:

````
Testing gtkextra
==> /usr/bin/clang -o test -D_REENTRANT -I/opt/X11/include/cairo -I/opt/X11/include/pixman-1 -I/opt/X11/include/libpng15 -I/opt/X11/include -I/opt/X11/
In file included from <built-in>:170:
<command line>:2:2: error: expected identifier or '('
 1
 ^
1 error generated.
Error: gtkextra: failed
Failed executing: /usr/bin/clang -o test -D_REENTRANT\ -I/opt/X11/include/cairo\ -I/opt/X11/include/pixman-1\ -I/opt/X11/include/libpng15\ -I/opt/X11/include\ -I/opt/X11/include/freetype2\ -I/usr/local/Cellar/gtkextra/3.1.2/include/gtkextra-3.0\ -I/usr/local/Cellar/gtk+/2.24.25/include/gtk-2.0\ -I/usr/local/Cellar/gtk+/2.24.25/lib/gtk-2.0/include\ -I/usr/local/Cellar/pango/1.36.8/include/pango-1.0\ -I/usr/local/Cellar/atk/2.14.0/include/atk-1.0\ -I/usr/local/Cellar/gdk-pixbuf/2.30.8/include/gdk-pixbuf-2.0\ -I/usr/local/Cellar/libpng/1.6.16/include/libpng16\ -I/usr/local/Cellar/pango/1.36.8/include/pango-1.0\ -I/usr/local/Cellar/harfbuzz/0.9.37/include/harfbuzz\ -I/usr/local/Cellar/pango/1.36.8/include/pango-1.0\ -I/usr/local/Cellar/glib/2.42.1/include/glib-2.0\ -I/usr/local/Cellar/glib/2.42.1/lib/glib-2.0/include\ -I/usr/local/opt/gettext/include\ 
 test.c -L/opt/X11/lib\ -L/usr/local/Cellar/gtkextra/3.1.2/lib\ -L/usr/local/Cellar/gtk+/2.24.25/lib\ -L/usr/local/Cellar/pango/1.36.8/lib\ -L/usr/local/Cellar/atk/2.14.0/lib\ -L/usr/local/Cellar/gdk-pixbuf/2.30.8/lib\ -L/usr/local/Cellar/glib/2.42.1/lib\ -L/usr/local/Cellar/pango/1.36.8/lib\ -L/usr/local/Cellar/glib/2.42.1/lib\ -L/usr/local/opt/gettext/lib\ -lgtkextra-x11-3.0\ -lgtk-x11-2.0\ -lgdk-x11-2.0\ -lpangocairo-1.0\ -latk-1.0\ -lcairo\ -lgdk_pixbuf-2.0\ -lgio-2.0\ -lpangoft2-1.0\ -lpango-1.0\ -lgobject-2.0\ -lglib-2.0\ -lintl\ -lfontconfig\ -lfreetype\ 
/usr/local/Library/Homebrew/formula.rb:703:in `block in system'
/usr/local/Library/Homebrew/formula.rb:660:in `open'
/usr/local/Library/Homebrew/formula.rb:660:in `system'
/usr/local/Library/Formula/gtkextra.rb:281:in `block in <class:Gtkextra>'
/usr/local/Library/Homebrew/formula.rb:613:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:21:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:610:in `run_test'
/usr/local/Library/Homebrew/cmd/test.rb:59:in `block (2 levels) in test'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/cmd/test.rb:58:in `block in test'
/usr/local/Library/Homebrew/cmd/test.rb:37:in `each'
/usr/local/Library/Homebrew/cmd/test.rb:37:in `test'
````

where exactly is `test.c` created? I would like to see its contents...

**Comment by mikemcquaid** [comment]
Ah, ok, fair enough. In that case use:
```ruby
libs = `pkg-config --libs`.chomp
system "some", "args", libs
```

**Comment by mikemcquaid** [comment]
`test.c` is created in a temporary directory which is deleted afterwards.

**Comment by tschoonj** [comment]
ok, things should work fine now...

**Comment by mikemcquaid** [comment]
Hmm, not finding a header unfortunately: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/19628/version=yosemite/testReport/junit/brew-test-bot/yosemite/test_gtkextra/

**Comment by tschoonj** [comment]
the buildbot needs explicit `pkg-config` access to the X11 headers. However this is not the case on my machine: the test runs fine here.

**Comment by tschoonj** [comment]
so even with an explicit dependency to :x11 things don't work on the buildbot. Can I somehow introduce the dependency within the test block itself?

**Comment by tschoonj** [comment]
I had a closer look at the error message. It only appears on the Yosemite and Mavericks buildbots for starters. The message is really weird too by the way:

````
        Testing gtkextra
Package xcb-shm was not found in the pkg-config search path.
Perhaps you should add the directory containing `xcb-shm.pc'
to the PKG_CONFIG_PATH environment variable
Package 'xcb-shm', required by 'cairo', not found
==> /usr/bin/clang -o test test.c -L/usr/local/Cellar/gtkextra/3.1.2/lib -L/usr/local/Cellar/gtk+/2.24.25/lib -L/usr/local/Cellar/pango/1.36.8/lib -L/usr/local/Cellar/atk/2.14.0/lib -L/usr/local/Cellar/cairo/1.14.0/lib -L/usr/local/Cellar/gdk-pixbuf/2.30.8/lib -L/usr/local/Cellar/glib/2.42.1/lib -L/usr/local/Cellar/pango/1.36.8/lib -L/usr/local/Cellar/glib/2.42.1/lib -L/usr/local/opt/gettext/lib -L/usr/local/Cellar/fontconfig/2.11.1/lib -L/usr/local/Cellar/freetype/2.5.5/lib -lgtkextra-x11-3.0 -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lintl -lfontconfig -lfreetype
test.c:1:10: fatal error: 'gtkextra/gtkextra.h' file not found
#include <gtkextra/gtkextra.h>
         ^
1 error generated.
````

If you look at the complete command here that was used to compile the testcase, it appears that the `pkg-config` command produced the correct output when it was invoked with the `--libs` flag (proven by the presence of the `-L` and `-l` flags), while it failed only when invoked with the `--cflags` flag.

I can't reproduce any of this on my Yosemite system by the way.

I am inclined to think that this is due to an issue on the buildbots, and not because of a problem in the formula... What do you think is going on here?

**Comment by mikemcquaid** [comment]
If it's an issue on the bots it'll still need fixed, I'm afraid.

**Comment by tschoonj** [comment]
Is there something else I could do?

By the way: this means that a test for Gtk+ or every other Formula that needs to use `pkg-config` will fail on the buildbot

**Comment by mikemcquaid** [comment]
It's not just an issue on the bots. I can reproduce the same thing running `brew test gtkextra` on my 10.10 machine.

**Comment by mikemcquaid** [comment]
```
Package xcb-shm was not found in the pkg-config search path.
Perhaps you should add the directory containing `xcb-shm.pc'
to the PKG_CONFIG_PATH environment variable
Package 'xcb-shm', required by 'cairo', not found
```

From the `test` output. So it doesn't actually use any cflags because `pkg-config --cflags gtkextra-3.0` produces the above error instead.

**Comment by tschoonj** [comment]
Interesting, I can reproduce it also on my machine when I run `unset PKG_CONFIG_PATH`, which I defined to include `/opt/X11/lib/pkgconfig/`

I assume that `depends_on :x11` (in this case put in the Gtk+ Formula), automatically adds this directory to PKG_CONFIG_PATH? Otherwise gtkextra would never survive the `configure`stage in the first place.

I guess that the solution involves getting `depends_on :x11` into the `test`block but that seems unsupported.

**Comment by mikemcquaid** [comment]
Try `ENV.prepend "PKG_CONFIG_PATH", OS::Mac::X11.lib/"pkg-config"` in the `test do`.

**Comment by tschoonj** [comment]
that seems to do the trick... pushing now and waiting for build-bot results

**Comment by tschoonj** [comment]
I am pretty sure `gtk_init` caused the examples to abort. Do the build-bots run an X server?

**Comment by mikemcquaid** [comment]
They do but we can slim down the example considerably. Literally all we need is to import a symbol from a header and link to make sure the include files and libraries are there. e.g. creating a GTKWidget and not doing anything with it and `return 0` is enough.

**Comment by tschoonj** [comment]
Something went wrong on Mountain Lion but I can't seem to find what exactly. All it says is that there are no failed tests...

File: Library/Formula/gtkextra.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    #include <gtkextra/gtkextra.h>
+    int main(int argc, char *argv[]) {
+      GtkWidget *canvas = gtk_plot_canvas_new(GTK_PLOT_A4_H, GTK_PLOT_A4_W, 0.8);
+      return 0;
+    }
+
+    EOS
+    cflags = `pkg-config --cflags gtkextra-3.0`.chomp.split
+    libs = `pkg-config --libs gtkextra-3.0`.chomp.split
+    system ENV.cc, "-o", "test", *cflags, "test.c", *libs
```
Comment (mikemcquaid) [inline Library/Formula/gtkextra.rb]: Can't use `*` more than once in Ruby 1.8 (on Mountain Lion). Instead do: `system ENV.cc, "-o", "test", "test.c", *(cflags + libs)

**Comment by tschoonj** [comment]
I noticed something weird: the trick with setting the `PKG_CONFIG_PATH` variable explicitly in the `test` block only works if I haven't set that variable myself in the shell environment. Otherwise I get the old missing header error.

Should I use something besides `prepend` to avoid this?

**Comment by mikemcquaid** [comment]
That shouldn't be the case and there's nothing I can think of that might avoid that.

**Comment by tschoonj** [comment]
everything seems to build nicely now... thanks for all your help!!!

**Comment by mikemcquaid** [comment]
Thanks for your patience and great work here!

**Comment by tschoonj** [comment]
Thanks for committing this Formula. You may want to report that issue with the `PKG_CONFIG_PATH` environment variable I encountered when running `brew test` to the Homebrew core developers: it looks fishy to me.

**Comment by mikemcquaid** [comment]
@tschoonj I am one of the core developers and I think it's expected, thanks.

**Comment by tschoonj** [comment]
Well I was surprised to see that `brew test`'s behavior depends on the user's environment variables. Especially so since `brew install` completely ignores them.

**Comment by mikemcquaid** [comment]
Yep, that's a known issue.

**Pull request closed without merging**